README.md
# CoralNet

[![CI Status](https://github.com/coralnet/coralnet/actions/workflows/django.yml/badge.svg)](https://github.com/coralnet/coralnet/actions/workflows/django.yml)

CoralNet is a website which serves as a repository and resource for benthic image analysis.

Website home: https://coralnet.ucsd.edu

Read more about us: https://coralnet.ucsd.edu/about/

CoralNet's machine-learning processes are powered by PySpacer: https://github.com/coralnet/pyspacer


## Building and viewing the documentation

You can browse the docs right in the GitHub repo under `/docs`, but GitHub doesn't support some of the ReStructured Text links. The build will get everything working: 

- Download or `git clone` this repository's code.
- Install Python and Sphinx. You can do this by either following the installation steps (`docs/installation.rst` in this repo) until Sphinx is installed, or you can use some other Python environment which already has Sphinx installed.
- Open a terminal/command line, cd to the `docs` directory, and run `make html`. (This command is cross platform, since there's a ``Makefile`` as well as a ``make.bat``.)
- Open `docs/_build/html/index.html` in a web browser to start browsing the documentation.
- It's also possible to output in formats other than HTML, if you use ``make <format>`` with a different format. See [Sphinx's docs](http://www.sphinx-doc.org/en/master/usage/quickstart.html#running-the-build).


CHANGELOG.md
# Changelog


For general instructions on how to update a development environment of CoralNet, see "Updating to the latest repository code" in `docs/server_operation.rst`. This changelog has specific instructions/notes for each CoralNet version.

For info about the semantic versioning used here, see `docs/versions.rst`.


## 1.7 (WIP)

- The new `EXTRACTORS_BUCKET` setting is now required when using a non-dummy feature extractor.
- The `MIN_NBR_ANNOTATED_IMAGES` setting has been renamed to `TRAINING_MIN_IMAGES`. Also, it's now tied to the corresponding pyspacer setting, which means that lowering this number will speed up unit tests.
- Updates to required packages:
  - pyspacer 0.4.1 -> 0.6.1
  - Pillow 9.4.0 -> 10.1.0
- New migrations to run for `annotations`, `events`, and `vision_backend`. annotations 0023 could possibly take hours per million images.

## [1.6](https://github.com/coralnet/coralnet/tree/1.6)

- Updates to required packages:
  - pyspacer 0.4.0 -> 0.4.1
  - numpy is now pinned to match pyspacer's requirements.txt
- New migrations to run for `jobs` and `vision_backend`.

## [1.5](https://github.com/coralnet/coralnet/tree/1.5)

- Python version has been updated from 3.6 to 3.10. See Server Operation > Upgrading Python in the CoralNet docs.
- Settings scheme has been changed. The old scheme used a developer-specific `.py` file AND a `secrets.json` file. The new scheme uses a `.env` file OR environment variables. Check the updated installation docs for details.
- Updates to required packages. Check requirements files for all the changes, but most notably:
  - Django 2.2.x -> 4.1.x
  - easy-thumbnails 2.6.0 -> our own fork
  - pyspacer 0.3.1 -> 0.4.0
- PostgreSQL version has been updated from 10 to 14. See Server Operation > Upgrading PostgreSQL in the CoralNet docs. CoralNet doesn't have any PostgreSQL-version-specific steps for this upgrade.
- New migrations to run for `api_core`, `calcification`, `labels`, `vision_backend`, Django's `auth`, and django-reversion's `reversion`.

### Notes

- A regression: unit tests now have 'noisy' log messages again, because the use of assertLogs() (replacing patch_logger() which was removed in Django 3.0) requires logging to be enabled. Ideally this regression would be fixed by reconfiguring (instead of disabling) the logging during tests, but that's something to figure out for a later release.
- Page header and footer nav-button styling has been cleaned up, so hopefully the shadowing makes a bit more sense now.

## [1.4](https://github.com/coralnet/coralnet/tree/1.4)

- Updates to required packages:
  - Removed `celery`.
  - Removed celery dependencies `amqp`, `anyjson`, `billiard`, and `kombu`.
  - Added `huey>=2.4.4,<2.5`.
  - Updated `redis==2.10.5` to `redis>=4.3.5,<4.4`.
- ``CELERY_ALWAYS_EAGER`` setting has been replaced with ``HUEY['immediate']``.
- Feel free to clean up any celery-generated files, such as the logs and schedule.

## [1.3](https://github.com/coralnet/coralnet/tree/1.3)

- Updates to required packages:
  - Removed `boto`.
  - Added `boto3==1.23.10`.

## [1.2](https://github.com/coralnet/coralnet/tree/1.2)

**Before updating from 1.1 to 1.2,** ensure that all spacer jobs and BatchJobs have finished first, because 1) spacer job token formats have changed for train and deploy, and 2) the migrations will be abandoning all existing BatchJobs. To do this:
  - Stop the web server and celery processes.
  - If using BatchQueue, check the AWS Batch console to ensure all your BatchJobs have finished. If using LocalQueue, then no waiting is needed here.
  - Manually run `collect_all_jobs()` in the shell to collect those jobs.

Once that's done, update to 1.2, then:

- Run migrations for ``api_core``, ``jobs`` (new app), and ``vision_backend``. Expect moderate runtime to process existing BatchJobs.

- Do an announcement after updating production. The new job-monitoring page for sources will hopefully answer many questions of the form "what's happening with the classifiers in my source?".

## [1.1](https://github.com/coralnet/coralnet/tree/1.1)

Update instructions have not been well tracked up to this point. If the migrations give you trouble, consider starting your environment fresh from 1.2 or later.

## [1.0](https://github.com/coralnet/coralnet/tree/1.0)

See [blog post](https://coralnet.ucsd.edu/blog/coralnet-is-officially-out-of-beta/) for major changes associated with this release.

settings.py
# Django settings for CoralNet project.
import settings_2
import os, sys
from django.conf.locale import en
import djcelery
djcelery.setup_loader()

abspath = lambda *p: os.path.abspath(os.path.join(*p))


PROJECT_ROOT = settings_2.PROJECT_ROOT

PROCESSING_ROOT = settings_2.PROCESSING_ROOT
TEST_PROCESSING_ROOT = settings_2.TEST_PROCESSING_ROOT

SLEEP_TIME_BETWEEN_IMAGE_PROCESSING = settings_2.SLEEP_TIME_BETWEEN_IMAGE_PROCESSING

USERENA_MODULE_PATH = abspath(PROJECT_ROOT, '..')
sys.path.insert(0, USERENA_MODULE_PATH)

DEBUG = settings_2.DEBUG

# TEMPLATE_DEBUG = True lets Sentry get template error info.  This
# won't reveal any error details to end users as long as DEBUG = False.
TEMPLATE_DEBUG = True

ADMINS = settings_2.ADMINS

MANAGERS = settings_2.MANAGERS

DATABASES = settings_2.DATABASES

# Default configurations to use when sending email
DEFAULT_FROM_EMAIL = settings_2.DEFAULT_FROM_EMAIL
SERVER_EMAIL = DEFAULT_FROM_EMAIL
#Default string to append in front of the subject of emails sent by the server
EMAIL_SUBJECT_PREFIX = '[CoralNet] '

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# On Unix systems, a value of None will cause Django to use the same
# timezone as the operating system.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/Los_Angeles'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

ugettext = lambda s: s
LANGUAGES = (
    ('en', ugettext('English')),
    ('nl', ugettext('Dutch')),
    ('fr', ugettext('French')),
    ('pl', ugettext('Polish')),
    ('pt', ugettext('Portugese')),
)

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale
USE_L10N = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/home/media/media.lawrence.com/media/"
# TODO: Might be better to move this directory out of the repository.
MEDIA_ROOT = abspath(PROJECT_ROOT, 'media')

# Media Root to be used during unit tests.
# This directory is best kept out of the repository.
TEST_MEDIA_ROOT = settings_2.TEST_MEDIA_ROOT

# Directory for sample files to be uploaded during unit tests.
# This directory should be in the repository.
SAMPLE_UPLOADABLES_ROOT = abspath(PROJECT_ROOT, 'sample_uploadables')

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://media.lawrence.com/media/", "http://example.com/media/"
MEDIA_URL = '/media/'

# File uploading
ORIGINAL_IMAGE_DIR = 'data/original/'
LABEL_THUMBNAIL_DIR = 'label_thumbnails/'
FILE_UPLOAD_MAX_MEMORY_SIZE = settings_2.FILE_UPLOAD_MAX_MEMORY_SIZE

# When uploading images and annotations together, the annotation dict needs
# to be kept on disk temporarily until all the Ajax upload requests are done.
# This is the directory where the dict files will be kept.
SHELVED_ANNOTATIONS_DIR = settings_2.SHELVED_ANNOTATIONS_DIR

DOCUMENT_ROOT = abspath(PROJECT_ROOT, 'docs')

STATIC_ROOT = settings_2.STATIC_ROOT

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# Additional locations of static files
# (besides apps' "static/" subdirectories, which are automatically included)
STATICFILES_DIRS = (
    # Put strings here, like "/home/chtml/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    # And remember to include a comma after the last element.

    # Project-wide static files
    abspath(PROJECT_ROOT, 'static'),
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
#    'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = settings_2.SECRET_KEY

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    'django.template.loaders.eggs.Loader',
)

TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.static",
    "django.contrib.messages.context_processors.messages",
    "django.core.context_processors.request",
)

# The order of middleware classes is important!
# https://docs.djangoproject.com/en/1.3/topics/http/middleware/
MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'userena.middleware.UserenaLocaleMiddleware',
    'sentry.client.middleware.Sentry404CatchMiddleware',

    # Put TransactionMiddleware after most non-cache middlewares that use the DB.
    # Any middleware specified after TransactionMiddleware will be
    # included in the same DB transaction that wraps the view function.
    # https://docs.djangoproject.com/en/1.3/topics/db/transactions/#tying-transactions-to-http-requests
    'django.middleware.transaction.TransactionMiddleware',

    # Put after TransactionMiddleware.
    # Model revisions needs to be part of transactions.
    'reversion.middleware.RevisionMiddleware',
)

ROOT_URLCONF = 'coralnet.urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    abspath(PROJECT_ROOT, 'templates')
)

# Define our project's installed apps separately from built-in and
# third-party installed apps.  This'll make it easier to define a
# custom test command that only runs our apps' tests.
# http://stackoverflow.com/a/2329425/859858
MY_INSTALLED_APPS = (
    'accounts',
    'images',
    'upload',
    'annotations',
    'visualization',
    'bug_reporting',
    'requests',
    'map',
    'lib',
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # Uncomment the next line to enable the admin:
    'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    'django.contrib.admindocs',
    'userena',
    'userena.contrib.umessages',
    'guardian',
    'easy_thumbnails',
    'south',
    'reversion',
    'sentry',
    'sentry.client',
    'djcelery',
    'GChartWrapper.charts',
	'djsupervisor',
)
# Add MY_INSTALLED_APPS to INSTALLED_APPS, each with a 'coralnet.' prefix.
INSTALLED_APPS += tuple(['coralnet.'+app_name for app_name in MY_INSTALLED_APPS])


# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        }
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}

AUTHENTICATION_BACKENDS = (
    'userena.backends.UserenaAuthenticationBackend',
    'guardian.backends.ObjectPermissionBackend',
    'django.contrib.auth.backends.ModelBackend', # this is default

)

TEST_RUNNER = 'lib.test_utils.MyTestSuiteRunner'

# This is our own setting (not a standard Django setting) specifying the
# verbosity of print messages printed by our unit tests' code. Note that
# this is different from Django's test runner's verbosity setting, which
# relates to messages printed by the test runner code.
UNIT_TEST_VERBOSITY = settings_2.UNIT_TEST_VERBOSITY

# Django-guardian settings
ANONYMOUS_USER_ID = -1
# For whatever reason, running tests in Postgres makes it complain when
# this ID is 0 or negative.
if 'test' in sys.argv or 'mytest' in sys.argv:
    ANONYMOUS_USER_ID = 99999999

# Other special user ids
IMPORTED_USER_ID = -2
ROBOT_USER_ID = -3
ALLEVIATE_USER_ID = -4

#Userena settings
LOGIN_REDIRECT_URL = '/images/source/'
LOGIN_URL = '/accounts/signin/'
LOGOUT_URL = '/accounts/signout/'
AUTH_PROFILE_MODULE = 'accounts.Profile'
USERENA_SIGNIN_REDIRECT_URL = LOGIN_REDIRECT_URL
USERENA_USE_MESSAGES = False
USERENA_LANGUAGE_FIELD = 'en'

# South settings: http://south.aeracode.org/docs/settings.html
SOUTH_MIGRATION_MODULES = {
    # Specify a nonexistent path like 'ignore' if an app has a migrations
    # folder in the default location and you want South to ignore it.

    # easy_thumbnails has South migrations, but they're not that well
    # maintained; the project's GitHub page has several open issues
    # related to South migrations (as of easy_thumbnails 1.0.3).
    'easy_thumbnails': 'ignore',
}

# Sentry settings: http://readthedocs.org/docs/sentry/en/latest/config/index.html
# SENTRY_TESTING enables usage of Sentry even when DEBUG = True
SENTRY_TESTING = True

# South settings
SOUTH_TESTS_MIGRATE = settings_2.SOUTH_TESTS_MIGRATE

#RabbitMQ hosts
BROKER_HOST = "localhost"
BROKER_PORT = 5672
BROKER_USER = settings_2.BROKER_USER
BROKER_PASSWORD = settings_2.BROKER_PASSWORD
BROKER_VHOST = settings_2.BROKER_VHOST

#Celery configuration
CELERYD_CONCURRENCY = settings_2.CELERYD_CONCURRENCY

# Google Maps
GOOGLE_MAPS_API_KEY = settings_2.GOOGLE_MAPS_API_KEY



# App URL bases
IMAGES_URL = '/images/'
ANNOTATIONS_URL = '/annotations/'
VISUALIZATION_URL = '/visualization/'
BUG_REPORTING_URL = '/feedback/'

# VISION BACKEND SETTINGS TODO: move to separate settings file.
NEW_MODEL_THRESHOLD = 1.5
MIN_NBR_ANNOTATED_IMAGES = 5
NBR_IMAGES_PER_LOOP = 100